### PR TITLE
Add `post_count` for free plan users in `add_post` api

### PR DESCRIPTION
**Pull request by RUTUKAPATE**
- `Subscription Check:` The function first retrieves the user's subscription type from the database.

- `Free Plan Verification:` Checks if the subscription type matches the free trial plan. 

- `Count Posts:` If the user is on the free plan, the function counts the number of posts associated with the user and returns this count.

- `Return None:` If the user is not on the free plan, the function returns None.

## Below `JSON` responses show the status `before` and `after` subscribing to the `Free Trial Plan`:

-  Before Subscribing to Free Trial

```json
{
    "message": "Please upgrade your subscription.",
    "success": false
}
```

- After Subscribing to Free Trial Plan and Adding First Post

```json
{
    "data": {
        "createdon": "2024-09-14 18:45:30",
        "mediaurl": "https://w0.peakpx.com/wallpaper/42/267/HD-wallpaper-nature-scenery-scenery-nature.jpg",
        "postcontent": "Purple beauty\\n\\nWhen the earth and the sky become one.",
        "postcount": 1,
        "postid": 22,
        "randompost": {
            "content": "",
            "date": "",
            "mediaurl": "",
            "time_ago": ""
        }
    },
    "message": "Post added successfully.",
    "success": true
}
```

### CODE REVIEW

File: backend/src/add_post.rs [hunk 1]
```diff
@@ -116,30 +116,28 @@ impl NewPost {
 }
 
 fn post_count_by_user_id(conn: &mut ft_sdk::Connection, user_id: i64) -> Option<i64> {
```
Comment (Arpita-Jaiswal) [inline backend/src/add_post.rs:118]: Write code documentation for this function.

File: backend/src/add_post.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+use diesel::ExpressionMethods;
+
```
Comment (Arpita-Jaiswal) [inline backend/src/add_post.rs:2]: Don't use `use`. Remove this line.

File: backend/src/add_post.rs [hunk 3]
```diff
@@ -90,26 +90,56 @@ impl NewPost {
...
             media_url: self.media_url,
             created_on: self.created_on.format("%Y-%m-%d %H:%M:%S").to_string(),
             random_post,
+            post_count,
         })
     }
 }
 
+/// Returns the number of posts by the user_id whose subscription is free trial
+fn post_count_by_user_id(conn: &mut ft_sdk::Connection, user_id: i64) -> Option<i64> {
```
Comment (amitu) [inline backend/src/add_post.rs:117]: The signature is wrong. You are eating errors up. If you ever do any operation that can return error (eg db is down etc), you must not eat the error. Change the return type to `Result<Option<i64>, PostCountError>`.

File: backend/src/add_post.rs [hunk 4]
```diff
@@ -90,26 +90,56 @@ impl NewPost {
...
                 None => PostWithTime {
                     content: "".to_string(),
                     media_url: "".to_string(),
                     time_ago: "".to_string(),
-                    date: "".to_string()
+                    date: "".to_string(),
                 },
             };
 
+        let post_count = post_count_by_user_id(conn, user_id)?;
```
Comment (amitu) [inline backend/src/add_post.rs:103]: Is the `?` converting `Option` to `Error` somehow here?

Comment (RUTUKAPATE) [reply backend/src/add_post.rs:103]: > Is the `?` converting `Option` to `Error` somehow here?

No, the `?` is not converting `Option` to `Error`. It only works with `Result` types. So, I `Added error handling` to propagate errors properly.

**Review by Arpita-Jaiswal — APPROVED** [review]

**Pull request merged**